src/ParseFasm.cpp

```diff
@@ -0,0 +1,574 @@
+       public:
+        virtual void syntaxError(Recognizer* recognizer,
+                                 Token* token,
+                                 size_t line,
+                                 size_t position,
+                                 const std::string& msg,
+                                 std::exception_ptr e) override {
+                throw ParseException{.line = line,
+                                     .position = position,
+                                     .message = std::string(msg)};
+        }
+};
+
+/// Common portion of 'from_string' and 'from_file'.
+/// Consumes an input stream and produces an output stream.
+static void parse_fasm(std::istream& in, std::ostream& out) {
+        ANTLRInputStream stream(in);
+        FasmLexer lexer(&stream);
+        FasmErrorListener errorListener;
+        lexer.removeErrorListeners();
+        lexer.addErrorListener(&errorListener);
+        CommonTokenStream tokens(&lexer);
+        FasmParser parser(&tokens);
+        parser.removeErrorListeners();
+        parser.addErrorListener(&errorListener);
+        auto* tree = parser.fasmFile();
+        FasmParserBaseVisitor(out).visit(tree);
+}
+
+/// Parse the given input string, returning output.
+/// Use hex mode (see above) if hex is true.
+/// Use a callback to avoid copying the result.
+void from_string(const char* in,
+                 bool hex,
+                 void (*ret)(const char* str, size_t),
+                 void (*err)(size_t, size_t, const char*)) {
+        hex_mode = hex;
+        std::istringstream input(in);
+        std::ostringstream output;
+
+        try {
+                parse_fasm(input, output);
+                output.put(0);
+                std::string result = output.str();
+                ret(result.c_str(), result.size());
+        } catch (ParseException e) {
+                // Parse failure will throw this exception.
+                err(e.line, e.position, e.message.c_str());
+        }
+}
+
+/// Parse the given input file, returning output.
+/// Use hex mode (see above) if hex is true.
+/// Use a callback to avoid copying the result.
+void from_file(const char* path,
+               bool hex,
+               void (*ret)(const char* str, size_t),
+               void (*err)(size_t, size_t, const char*)) {
+        hex_mode = hex;
+        std::fstream input(std::string(path), input.in);
+        std::ostringstream output;
+        if (input.is_open()) {
+                try {
+                        parse_fasm(input, output);
+                        output.put(0);
+                        std::string result = output.str();
+                        ret(result.c_str(), result.size());
+                } catch (ParseException e) {
+                        // Parse failure will throw this exception.
+                        err(e.line, e.position, e.message.c_str());
+                }
+        } else {
+                err(0, 0, "Couldn't open file");
+        }
+}
+
+#ifndef TEST
```

Comment:
This should probably be a separate `.cpp` rather than using an `ifndef` like this.

Reply:
I didn't want to create a header full of boilerplate just for a test driver.

Reply:
I've moved the code into `ParseFasmRun.cpp`.